src/main/java/pl/agh/edu/iisg/io/vmms/vmmsbackend/controller/VMPoolController.java

```diff
@@ -43,46 +43,113 @@ public VMPoolController(VMPoolService vmPoolService) {
-                            }else if(e.getMessage().contains("constraint [fk_")){
-                                details = " - Foreign key";
-                            }
-                            throw new VMPoolImportFileException("Constraint not met" + details);
+            vmPoolsNumber = updateVmPools(file, vmPoolsNumber, true);
+        }
+
+        return "Created " + vmPoolsNumber + " virtual machine pools.";
+    }
+
+    private int updateVmPools(@RequestParam("file") MultipartFile file, int vmPoolsNumber, boolean replace) throws VMPoolImportFileException {
+        try {
+            String line;
+            BufferedReader bufferedReader = new BufferedReader(
+                    new InputStreamReader(file.getInputStream()));
+            VMPoolCSVParser parser = new VMPoolCSVParser();
+
+            List<VMPool> vmPools = vmPoolService.getVMPools();
+
+            StringBuilder errors = new StringBuilder();
+            int lines = 0;
+
+            while ((line = bufferedReader.readLine()) != null) {
+                lines++;
+                Optional<VMPool> o = parser.parseLine(line);
+                if (o.isPresent()) {
+                    VMPool newPool = o.get();
+                    VMPool found = vmPoolService.find(newPool.getShortName());
+                    try {
+                        if (processVmPool(vmPools, newPool, found)) {
+                            vmPoolsNumber++;
                         }
-                        vmPoolsNumber++;
-                    }
-                    else {
-                        throw new VMPoolImportFileException("Failed to parse the file!");
+                    } catch (VMPoolImportFileException e){
+                        errors.append("Error in line ").append(lines).append(": ").append(e.getMessage()).append("\n");
                     }
+                } else {
+                    errors.append("Error parsing line ").append(lines).append("\n");
                 }
+            }
 
-            } catch (IOException e) {
-                throw new VMPoolImportFileException("Failed to open the file!");
+            if (replace) {
+                for (VMPool deleted : vmPools) {
+                    if (deleted.getReservations().size() > 0) {
+                        errors.append("Can not delete pool ").append(deleted.getDisplayName()).append(" with trailing reservations!\n");
+                    }
+                    vmPoolService.delete(deleted);
+                }
             }
+            if(errors.length() != 0){
+                throw new VMPoolImportFileException(errors.toString());
+            }
+
+        } catch (IOException e) {
+            throw new VMPoolImportFileException("Failed to open the file!");
         }
+        return vmPoolsNumber;
+    }
 
-        return "Created " + vmPoolsNumber + " virtual machine pools.";
+    private boolean processVmPool(List<VMPool> vmPools, VMPool newPool, VMPool found) throws VMPoolImportFileException {
```

Comment:
The found parameter is cryptic. I think it would be better to perform the search inside this method and remove the parameter.